src/eval/lazy.rs

```diff
@@ -121,41 +124,149 @@ impl ThunkData {
+                    body: as_function,
+                    env,
+                })
+            }
+        }
+    }
+
     /// Return a reference to the closure currently cached.
     pub fn closure(&self) -> &Closure {
         match self.inner {
             InnerThunkData::Standard(ref closure) => closure,
-            InnerThunkData::Revertible { ref cached, .. } => cached,
+            // Nothing should peek into a revertible thunk before the cached value has been
+            // constructed by [`build_cached_value`]. This is an invariant that MUST be maintained
+            // by the interpreter.
+            //
+            // `cached` set to `None` solely exists because we need to first allocate all the
+            // revertible thunks corresponding to a recursive record, and only then can we patch
+            // them (build the cached value) in a second step. But calling to
+            // [`ThunkData::new_rev`] followed by [`ThunkData::build_cached_value`] should be logically
+            // seen as just one construction operation.
+            InnerThunkData::Revertible { ref cached, .. } => cached
+                .as_ref()
+                .expect("tried to get data from a revertible thunk without a cached value"),
         }
     }
 
     /// Return a mutable reference to the closure currently cached.
     pub fn closure_mut(&mut self) -> &mut Closure {
         match self.inner {
             InnerThunkData::Standard(ref mut closure) => closure,
-            InnerThunkData::Revertible { ref mut cached, .. } => Rc::make_mut(cached),
+            InnerThunkData::Revertible {
+                ref mut cached,
+                ref mut orig,
+                ..
+            } => cached.as_mut().unwrap_or_else(|| Rc::make_mut(orig)),
         }
     }
 
     /// Consume the data and return the cached closure.
     pub fn into_closure(self) -> Closure {
         match self.inner {
             InnerThunkData::Standard(closure) => closure,
-            InnerThunkData::Revertible { orig, cached, .. } => {
-                std::mem::drop(orig);
-                Rc::try_unwrap(cached).unwrap_or_else(|rc| (*rc).clone())
+            // Nothing should access the cached value of a revertible thunk before the cached
+            // value has been constructed. This is an invariant that MUST be maintained by the
+            // interpreter
+            //
+            // `cached` set to `None` solely exists because we need to first allocate all the
+            // revertible thunks corresponding to a recursive record, and only then can we patch
+            // them (build the cached value) in a second step. But calling to
+            // [`ThunkData::new_rev`] followed by [`ThunkData::build_cached_value`] should be logically
+            // seen as just one construction operation.
+            InnerThunkData::Revertible { cached, .. } => {
+                cached.expect("tried to get data from a revertible thunk without a cached value")
```

Comment:
Are you talking about the comment or the error message inside `expect` ? I imagine you meant the latter, in which case I agree.

Reply:
I actually meant _both_. I'm suggesting combining them so that the comment _is_ the error message passed to expect. I'm not sure it's a perfect solution, but I can definitely see a future where the comment gets changed in one place but gets missed in others. 

That said, I completely understand if you decide not to do this, as it does mean that devs need to go look up the static string definition to find the info that's currently in the comment...

Reply:
Mmh, I think I like the idea of the comment being in the right place, directly accessible. The text showed in `expect` doesn't have exactly the same purpose. I'll still extract it in a separate variable though